### DISCUSSION TITLE
Lazy initialization of variables

### DISCUSSION THREAD

**Discussion by smartiniOnGitHub** [vlang/v#11791] [Ideas]
Hi all,
**for the future**, what do you think on the ability to implement "lazy initialization variables", so initial value will be deferred to its first usage ?
Not sure if that would be useful to constants.
Maybe implemented with some compiler magic (`$lazy` ?) or a new keyword (`lazy`) or an attribute ...

Some reference:
- https://en.wikipedia.org/wiki/Lazy_initialization
- https://dotty.epfl.ch/docs/reference/changed-features/lazy-vals-init.html
- https://dzone.com/articles/scala-lazy-evaluation
- https://www.baeldung.com/scala/lazy-val

In Scala it has been implemented time ago and it's useful (at least in some cases) with the addition of the keyword `lazy` (but only for variables); but I'd not take in account the ability (like in Scala) to initialize with a code block, to avoid too complex stuff here.

What do you think ?

Thanks a lot.

**Comment by JalonSolov** [comment]
Definitely not needed for constants, as those are initialized at compile time, before the executable is even created (or as it's being created...).

Personally, I prefer to solve these issues myself, without extra keywords.  Either that, or have the compiler figure it out for me as an optimization.

Basically, if you know enough to add the lazy keyword, why not restructure your code so the initialization only happens when it needs to anyway?  The compiler optimization could catch small things you missed.

**Comment by ntrel** [comment]
> those are initialized at compile time, before the executable is even created (or as it's being created.

Last time I checked V constants were initialized at program start up (at least when calling a function), so they are just immutable global variables. So if a constant isn't necessarily used and initialization is slow, lazy initialization for `const` could be useful.

> why not restructure your code so the initialization only happens when it needs to anyway

If the variable is part of a library API, you can't update user's code (or even see it necessarily).

**Comment by JalonSolov** [comment]
Ah, yes, functions to initialize constants.  Ugh.  I would just make them variables (perhaps in a struct) and not worry about it.

True, you can't see their stuff without examining their source, but that's where making this a compiler optimization step would solve the problem for you.

**Comment by smartiniOnGitHub** [comment]
Hi all and thanks for the feedback; not sure all cases could be covered only by compiler optimizations, but of course I understand that for now keep things simple and stable is a major goal.
Just for info, in Scala lazy variables are useful to simplify/interact with infinite data structures (like Stream, etc) and if I remember well even with some async field (in future V maybe even struct fields via an attribute); but we can discuss later if/when needed ...
Thanks for now. Bye
